src/java.base/share/classes/java/util/concurrent/DelayScheduler.java

```diff
@@ -0,0 +1,554 @@
+     * policy if not already done so, and if not empty or pool not
+     * terminating, returns.  Otherwise, cancels all tasks in heap and
+     * pending queue.
+     * @return negative if stop, else current heap size.
+     */
+    private int tryStop(ForkJoinPool p, ScheduledForkJoinTask<?>[] h, int n,
+                        int runStatus, int prevRunStatus) {
+        if ((runStatus & POOL_STOPPING) == 0) {
+            if (n > 0) {
+                if (cancelDelayedTasksOnShutdown != 0) {
+                    cancelAll(h, n);
+                    n = 0;
+                }
+                else if (prevRunStatus == 0 && h != null && h.length >= n) {
+                    ScheduledForkJoinTask<?> t; int stat; // remove periodic tasks
+                    for (int i = n - 1; i >= 0; --i) {
+                        if ((t = h[i]) != null &&
+                            ((stat = t.status) < 0 || t.nextDelay != 0L)) {
+                            t.heapIndex = -1;
+                            if (stat >= 0)
+                                t.trySetCancelled();
+                            n = replace(h, i, n);
+                        }
+                    }
+                }
+            }
+            if (n > 0 || p == null || !p.tryStopIfShutdown(this))
+                return n;       // check for quiescent shutdown
+        }
+        if (n > 0)
+            cancelAll(h, n);
+        for (ScheduledForkJoinTask<?> a = (ScheduledForkJoinTask<?>)
+                 U.getAndSetReference(this, PENDING, null);
+             a != null; a = a.nextPending)
+            a.trySetCancelled(); // clear pending requests
+        return -1;
+    }
+
+    private static void cancelAll(ScheduledForkJoinTask<?>[] h, int n) {
+        if (h != null && h.length >= n) {
+            ScheduledForkJoinTask<?> t;
+            for (int i = 0; i < n; ++i) {
+                if ((t = h[i]) != null) {
+                    h[i] = null;
+                    t.heapIndex = -1;
+                    t.trySetCancelled();
+                }
+            }
+        }
+    }
+
+    /**
+     * Task class for DelayScheduler operations
+     */
+    @SuppressWarnings("serial")    // Not designed to be serializable
+    static final class ScheduledForkJoinTask<T>
+        extends ForkJoinTask.InterruptibleTask<T>
+        implements ScheduledFuture<T> {
+        ForkJoinPool pool;         // nulled out after use
+        Runnable runnable;         // at most one of runnable or callable nonnull
+        Callable<? extends T> callable;
+        T result;
+        ScheduledForkJoinTask<?> nextPending; // for DelayScheduler pending queue
+        long when;                  // nanoTime-based trigger time
+        final long nextDelay;       // 0: once; <0: fixedDelay; >0: fixedRate
+        int heapIndex;              // if non-negative, index on heap
+        final boolean isImmediate;  // run by scheduler vs submitted when ready
```

Comment:
If desired, it would be possible to create a bitset for `isImmediate` and `isCallable` such that we could get rid of a reference field per ScheduledForkJoinTask.
I suspect it would also be possible to use the same field for `pool` and `result` as there shouldn't be a result available for as long as `pool` is non-null.

Might not be worth doing at this point, but saving 2 reference fields for something which might get allocated millions of times might be worth it?